Data/ByteString/Internal/Type.hs

```diff
@@ -354,9 +354,17 @@ instance IsString ByteString where
 
 instance Data ByteString where
   gfoldl f z txt = z packBytes `f` unpackBytes txt
-  toConstr _     = error "Data.ByteString.ByteString.toConstr"
-  gunfold _ _    = error "Data.ByteString.ByteString.gunfold"
-  dataTypeOf _   = mkNoRepType "Data.ByteString.ByteString"
+  toConstr _     = packConstr
+  gunfold k z c = case constrIndex c of
+    1 -> k (z packBytes)
+    _ -> error "gunfold: unexpected constructor of strict ByteString"
+  dataTypeOf _   = byteStringDataType
```

Comment:
The `dataTypeOf` change affects non-throwing traces. Major version bump?

Reply:
PVP does not prescribe this: somewhat surprisingly, it does not say anything about term-level changes. I deem this as a bug fix, not a change of _intended_ semantics.

Reply:
The intent of the existing instance implementation is not so obvious. I do think this `dataTypeOf` change makes the instance more self-consistent and is an improvement.

But as a non-expert in the wider usage of `Data` I'm hesitant to put even small potentially-breaking changes into a minor release. And our users seem in no particular hurry to get working `gread` functionality.

The rest of the patch is fine for a minor release, of course.

> PVP does not prescribe this: somewhat surprisingly, it does not say anything about term-level changes.

Right, the _text_ of PVP mentions nothing on this subject. I'm reasonably sure that's an oversight. The [graphical decision tree](https://pvp.haskell.org/#decision-tree) on the front page does have a node indicating that if "the behavior of any exported function change[d]," then a major version bump is necessary. The [FAQ also provides an explicit example](https://pvp.haskell.org/faq/#what-are-some-real-world-examples-of-packages-causing-breakage-due-to-semantic-changes) on this subject.

Reply:
> I'm reasonably sure that's an oversight. 

Neither FAQ nor decision tree (https://github.com/haskell/pvp/pull/52) are normative. PVP is compile-time only. There was an attempt to change this (https://github.com/haskell/pvp/issues/30), but it was withdrawn. 

It's not a hill I want to die on, so if you feel strongly about it, I'll revert changes to `dataTypeOf`.

Reply:
No need to revert! The change is fine; I just think this should be where `bytestring-0.12` finally diverges from `master`.